File: 502. IPO/Solution.py
import heapq

class Solution:
    def findMaximizedCapital(self, k: int, w: int, profits: List[int], capital: List[int]) -> int:
        n, pointer = len(profits), 0
        combined, pq = sorted(list(zip(capital, profits)), key = itemgetter(0)), []
        for i in range(k):
            while pointer < n and combined[pointer][0] <= w:
                heapq.heappush(pq, -combined[pointer][1])
                pointer += 1
            if not pq: break
            w -= heapq.heappop(pq)
        return w

File: 502. IPO/Solution.java
class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        int n = profits.length, pointer = 0;
        int[][] combined = new int[n][];
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);
        for (int i = 0; i < n; i++) combined[i] = new int[] { capital[i], profits[i] };
        Arrays.sort(combined, (a, b) -> a[0] - b[0]);
        for (int i = 0; i < k; i++) {
            while (pointer < n && combined[pointer][0] <= w) {
                pq.offer(combined[pointer][1]);
                pointer++;
            }
            if (pq.isEmpty()) break;
            w += pq.poll();
        }
        return w;
    }
}